wire-library/wire-runtime/src/jvmMain/kotlin/com/squareup/wire/DurationProtoAdapter.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire
+
+import com.squareup.wire.FieldEncoding.LENGTH_DELIMITED
+import java.time.Duration
+
+internal object DurationProtoAdapter : ProtoAdapter<Duration>(
+    fieldEncoding = LENGTH_DELIMITED,
+    type = Duration::class,
+    typeUrl = "type.googleapis.com/google.protobuf.Duration"
+) {
+  override fun encodedSize(value: Duration): Int {
+    var result = 0
+    val seconds = value.sameSignSeconds
+    if (seconds != 0L) result += INT64.encodedSizeWithTag(1, seconds)
+    val nanos = value.sameSignNanos
+    if (nanos != 0) result += INT32.encodedSizeWithTag(2, nanos)
+    return result
+  }
+
+  override fun encode(writer: ProtoWriter, value: Duration) {
+    val seconds = value.sameSignSeconds
+    if (seconds != 0L) INT64.encodeWithTag(writer, 1, seconds)
+    val nanos = value.sameSignNanos
+    if (nanos != 0) INT32.encodeWithTag(writer, 2, nanos)
+  }
+
+  override fun decode(reader: ProtoReader): Duration {
+    var seconds = 0L
+    var nanos = 0
+    reader.forEachTag { tag ->
+      when (tag) {
+        1 -> seconds = INT64.decode(reader)
+        2 -> nanos = INT32.decode(reader)
+        else -> reader.readUnknownField(tag)
+      }
+    }
+    return Duration.ofSeconds(seconds, nanos.toLong())
```

Comment:
The proto defines that the nanosecond offset is signed whereas the Duration type wants it to be unsigned. This needs to handle the case where the value is negative by subtracting 1 from the seconds and flipping the nanos to be positive.

This is also a problem for the common proto adapter.

Reply:
The Java class does it automatically.